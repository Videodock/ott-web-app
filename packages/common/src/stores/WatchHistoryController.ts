import { injectable } from 'inversify';

import WatchHistoryService from '../services/WatchHistoryService';
import type { PlaylistItem } from '../../types/playlist';
import type { WatchHistoryItem } from '../../types/watchHistory';

import { useAccountStore } from './AccountStore';
import { useConfigStore } from './ConfigStore';
import { useWatchHistoryStore } from './WatchHistoryStore';

@injectable()
export default class WatchHistoryController {
  private readonly watchHistoryService: WatchHistoryService;

  constructor(watchHistoryService: WatchHistoryService) {
    this.watchHistoryService = watchHistoryService;
  }

  initialize = async () => {
    await this.restoreWatchHistory();
  };

  restoreWatchHistory = async () => {
    const { user } = useAccountStore.getState();
    const continueWatchingList = useConfigStore.getState().config.features?.continueWatchingList;

    if (!continueWatchingList) {
      return;
    }

    const watchHistory = await this.watchHistoryService.getWatchHistory(user, continueWatchingList);

    useWatchHistoryStore.setState({
      watchHistory: watchHistory.filter((item): item is WatchHistoryItem => !!item?.mediaid),
      playlistItemsLoaded: true,
      continueWatchingPlaylistId: continueWatchingList,
    });
  };

  persistWatchHistory = async () => {
    const { watchHistory } = useWatchHistoryStore.getState();
    const { user } = useAccountStore.getState();

    await this.watchHistoryService.persistWatchHistory(user, watchHistory);
  };

  /**
   *  If we already have an element with continue watching state, we:
   *    1. Update the progress
   *    2. Move the element to the continue watching list start
   *  Otherwise:
   *    1. Move the element to the continue watching list start
   *    2. If there are many elements in continue watching state we remove the oldest one
   */
  saveItem = async (item: PlaylistItem, seriesItem: PlaylistItem | undefined, videoProgress: number | null) => {
    const { watchHistory } = useWatchHistoryStore.getState();

    if (!videoProgress) return;

    const updatedHistory = await this.watchHistoryService.saveItem(item, seriesItem, videoProgress, watchHistory);

    if (updatedHistory) {
      useWatchHistoryStore.setState({ watchHistory: updatedHistory });
      await this.persistWatchHistory();
    }
  };
}
