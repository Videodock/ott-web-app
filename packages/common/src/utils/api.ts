import type { CommonResponse } from '@inplayer-org/inplayer.js';

import type { InPlayerError } from '../../types/inplayer';

export class ApiError extends Error {
  code: number;
  message: string;

  constructor(message = '', code: number) {
    super(message);
    this.name = 'ApiError';
    this.message = message;
    this.code = code;
  }
}

/**
 * Get data
 * @param response
 */
export const getDataOrThrow = async (response: Response) => {
  const data = await response.json();

  if (!response.ok) {
    const message = `Request '${response.url}' failed with ${response.status}`;
    const apiMessage = data && typeof data === 'object' && 'message' in data && typeof data.message === 'string' ? data.message : undefined;

    throw new ApiError(apiMessage || message, response.status || 500);
  }

  return data;
};

export const getCommonResponseData = (response: { data: CommonResponse }) => {
  const { code, message } = response.data;
  if (code !== 200) {
    throw new Error(message);
  }
  return {
    errors: [],
    responseData: {
      message,
      code,
    },
  };
};

export const isCommonError = (error: unknown): error is InPlayerError => {
  return (
    typeof error === 'object' &&
    error !== null &&
    'response' in error &&
    typeof (error as InPlayerError).response?.data?.code === 'number' &&
    typeof (error as InPlayerError).response?.data?.message === 'string'
  );
};
