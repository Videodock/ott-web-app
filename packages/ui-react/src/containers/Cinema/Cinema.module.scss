@use '@jwp/ott-ui-react/src/styles/variables';
@use '@jwp/ott-ui-react/src/styles/theme';
@use '@jwp/ott-ui-react/src/styles/mixins/responsive';
@use '@jwp/ott-ui-react/src/styles/mixins/typography';

.cinemaContainer {
  width: 100%;
  height: 100%;
}

.cinema {
  width: 100vw;
  height: calc(100vh - calc(100vh - 100%));
  overflow: hidden;
  background-color: variables.$black;
}

.playerOverlay {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1;
  width: 100%;
  min-height: 100px;
  padding: 24px 56px;
  background-image: linear-gradient(to top, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 0.1) 25%, rgba(0, 0, 0, 0.25) 100%);
  pointer-events: none;

  @include responsive.mobile-only() {
    padding-left: 16px;
  }
}

.playerContent {
  display: flex;
  width: 80%;

  @include responsive.mobile-only() {
    width: 100%;
  }
}

.backButton {
  margin-right: 24px;
  opacity: 1;
  pointer-events: auto;

  > svg {
    width: 36px;
    height: 36px;
    -webkit-filter: drop-shadow(2px 2px 5px rgba(0, 0, 0, 0.4)); /* stylelint-disable-line */
    filter: drop-shadow(2px 2px 5px rgba(0, 0, 0, 0.4));
  }
}

.title {
  @include typography.video-title-base;
}

.metaContainer {
  display: flex;
  flex-direction: column;
}

.primaryMetadata {
  display: flex;
  align-items: center;
  margin-bottom: 8px;
  font-size: 16px;
  line-height: variables.$base-line-height;
  letter-spacing: 0.15px;

  @include responsive.mobile-only() {
    font-size: 14px;
  }
}

.secondaryMetadata {
  margin-top: 4px;
  margin-bottom: 8px;
  font-size: 20px;
  line-height: variables.$base-line-height;
  letter-spacing: 0.5px;

  @include responsive.mobile-only() {
    order: 1;
    margin: 4px 0;
    font-size: 18px;
  }
}
