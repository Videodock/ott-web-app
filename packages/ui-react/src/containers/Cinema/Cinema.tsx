import React, { useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import type { PlaylistItem } from '@jwp/ott-common/types/playlist';
import ArrowLeft from '@jwp/ott-theme/assets/icons/arrow_left.svg?react';

import IconButton from '../../components/IconButton/IconButton';
import PlayerContainer from '../PlayerContainer/PlayerContainer';
import Fade from '../../components/Animation/Fade/Fade';
import Icon from '../../components/Icon/Icon';
import Modal from '../../components/Modal/Modal';

import styles from './Cinema.module.scss';

type Props = {
  open: boolean;
  item: PlaylistItem;
  seriesItem?: PlaylistItem;
  onPlay?: () => void;
  onPause?: () => void;
  onComplete?: () => void;
  onClose?: () => void;
  onNext?: () => void;
  feedId?: string;
  title: string;
  primaryMetadata: React.ReactNode;
  secondaryMetadata?: React.ReactNode;
  liveStartDateTime?: string | null;
  liveEndDateTime?: string | null;
  liveFromBeginning?: boolean;
};

const Cinema: React.FC<Props> = ({
  open,
  item,
  seriesItem,
  title,
  primaryMetadata,
  secondaryMetadata,
  onPlay,
  onPause,
  onComplete,
  onClose,
  onNext,
  feedId,
  liveStartDateTime,
  liveEndDateTime,
  liveFromBeginning,
}: Props) => {
  const { t } = useTranslation();

  // state
  const [isPlaying, setIsPlaying] = useState<boolean>(false);
  const [userActive, setUserActive] = useState(true);

  const handlePlay = useCallback(() => {
    setIsPlaying(true);
    onPlay && onPlay();
  }, [onPlay]);

  const handlePause = useCallback(() => {
    setIsPlaying(false);
    onPause && onPause();
  }, [onPause]);

  const handleComplete = useCallback(() => {
    onComplete && onComplete();
  }, [onComplete]);

  const handleNext = useCallback(() => {
    onNext && onNext();
  }, [onNext]);

  const handleUserActive = useCallback(() => setUserActive(true), []);
  const handleUserInactive = useCallback(() => setUserActive(false), []);

  // effects
  useEffect(() => {
    if (open) setUserActive(true);
  }, [open]);

  return (
    <Modal open={open} animationContainerClassName={styles.cinemaContainer} onClose={onClose}>
      <div className={styles.cinema}>
        <Fade open={!isPlaying || userActive} keepMounted>
          <div className={styles.playerOverlay}>
            <div className={styles.playerContent}>
              <IconButton aria-label={t('common:back')} onClick={onClose} className={styles.backButton}>
                <Icon icon={ArrowLeft} />
              </IconButton>
              <div>
                <h1 className={styles.title}>{title}</h1>
                <div className={styles.metaContainer}>
                  {secondaryMetadata && <div className={styles.secondaryMetadata}>{secondaryMetadata}</div>}
                  <div className={styles.primaryMetadata}>{primaryMetadata}</div>
                </div>
              </div>
            </div>
          </div>
        </Fade>
        <PlayerContainer
          item={item}
          seriesItem={seriesItem}
          feedId={feedId}
          autostart={true}
          onPlay={handlePlay}
          onPause={handlePause}
          onComplete={handleComplete}
          onUserActive={handleUserActive}
          onUserInActive={handleUserInactive}
          onNext={handleNext}
          liveEndDateTime={liveEndDateTime}
          liveFromBeginning={liveFromBeginning}
          liveStartDateTime={liveStartDateTime}
        />
      </div>
    </Modal>
  );
};

export default Cinema;
