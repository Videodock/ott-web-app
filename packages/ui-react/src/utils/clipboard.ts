export const copyToClipboard = (value: string): void => {
  const inputElement = document.createElement('input');
  inputElement.style.zIndex = '-10';
  inputElement.style.position = 'absolute';
  inputElement.style.left = '-1000';
  inputElement.value = value;
  document.body.appendChild(inputElement);
  inputElement.select();
  document.execCommand('copy');
  inputElement.blur();
  document.body.removeChild(inputElement);
};
