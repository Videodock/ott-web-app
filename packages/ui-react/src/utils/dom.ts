export const canUseDOM = !!(typeof window !== 'undefined' && window.document && window.document.createElement);

let size: number;

export default function scrollbarSize(recalc?: boolean) {
  if ((!size && size !== 0) || recalc) {
    if (canUseDOM) {
      const scrollDiv = document.createElement('div');
      scrollDiv.style.position = 'absolute';
      scrollDiv.style.top = '-9999px';
      scrollDiv.style.width = '50px';
      scrollDiv.style.height = '50px';
      scrollDiv.style.overflow = 'scroll';
      document.body.appendChild(scrollDiv);
      size = scrollDiv.offsetWidth - scrollDiv.clientWidth;
      document.body.removeChild(scrollDiv);
    }
  }
  return size;
}

export const addScript = (src: string): Promise<void> => {
  return new Promise((resolve) => {
    const script: HTMLScriptElement = document.createElement('script');
    script.type = 'text/javascript';
    script.async = true;
    script.src = src;
    script.onload = () => resolve();
    script.onerror = (error) => {
      console.info('Error loading external script', error);
      throw new Error('Error loading external script');
    };
    document.body.appendChild(script);
  });
};
