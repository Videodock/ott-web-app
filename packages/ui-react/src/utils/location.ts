import { createURL, type QueryParamsArg } from '@jwp/ott-common/src/utils/urlFormatting';
import type { Location } from 'react-router-dom';

import type { AccountModals } from '../containers/AccountModal/AccountModal';

export const createURLFromLocation = (location: Location, queryParams: QueryParamsArg = {}) => {
  return createURL(`${location.pathname}${location.search || ''}`, queryParams);
};

export const modalURLFromLocation = (location: Location, u: keyof AccountModals | null, queryParams?: QueryParamsArg) => {
  return createURL(`${location.pathname}${location.search || ''}`, { u, ...queryParams });
};
