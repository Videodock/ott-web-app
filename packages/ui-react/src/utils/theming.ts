import type { Config } from '@jwp/ott-common/types/config';
import { calculateContrastColor } from '@jwp/ott-common/src/utils/common';
import env from '@jwp/ott-common/src/env';

export const setThemingVariables = (config: Config) => {
  const root = document.querySelector(':root') as HTMLElement;
  const { backgroundColor, highlightColor, headerBackground } = config.styling || {};
  const bodyFont = env.APP_BODY_FONT;
  const bodyAltFont = env.APP_BODY_ALT_FONT;

  if (!root) return;

  if (backgroundColor) {
    root.style.setProperty('--body-background-color', backgroundColor);
    root.style.setProperty('--background-contrast-color', calculateContrastColor(backgroundColor));
  }
  if (highlightColor) {
    root.style.setProperty('--highlight-color', highlightColor);
    root.style.setProperty('--highlight-contrast-color', calculateContrastColor(highlightColor));
  }
  if (headerBackground) {
    root.style.setProperty('--header-background', headerBackground);
    root.style.setProperty('--header-contrast-color', calculateContrastColor(headerBackground));
  }

  if (bodyFont) {
    root.style.setProperty('--body-font-family', bodyFont);
  }
  if (bodyAltFont) {
    root.style.setProperty('--body-alt-font-family', bodyAltFont);
  }
};
