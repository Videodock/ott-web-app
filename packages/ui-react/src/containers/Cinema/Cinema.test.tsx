import React from 'react';
import type { PlaylistItem } from '@jwp/ott-common/types/playlist';
import { beforeEach } from 'vitest';
import { mockService } from '@jwp/ott-common/test/mockService';
import ApiService from '@jwp/ott-common/src/services/ApiService';
import GenericEntitlementService from '@jwp/ott-common/src/services/GenericEntitlementService';
import JWPEntitlementService from '@jwp/ott-common/src/services/JWPEntitlementService';
import WatchHistoryController from '@jwp/ott-common/src/controllers/WatchHistoryController';
import { act } from '@testing-library/react';

import { renderWithRouter } from '../../../test/utils';

import Cinema from './Cinema';

describe('<Cinema>', () => {
  beforeEach(() => {
    mockService(ApiService, {});
    mockService(GenericEntitlementService, {});
    mockService(JWPEntitlementService, {});
    mockService(WatchHistoryController, {});
  });

  test('renders and matches snapshot', async () => {
    const item = {
      description: 'Test item description',
      duration: 354,
      feedid: 'ax85aa',
      image: 'http://test/img.jpg',
      images: [],
      link: 'http://test/link',
      genre: 'Tester',
      mediaid: 'zp50pz',
      pubdate: 26092021,
      rating: 'CC_CC',
      sources: [],
      seriesId: 'ag94ag',
      tags: 'Test tag',
      title: 'Test item title',
      tracks: [],
    } as PlaylistItem;

    const { baseElement } = await act(() =>
      renderWithRouter(
        <Cinema
          item={item}
          onPlay={() => null}
          onPause={() => null}
          open
          title={item.title}
          primaryMetadata="Primary metadata"
          onClose={vi.fn()}
          onNext={vi.fn()}
        />,
      ),
    );

    expect(baseElement).toMatchSnapshot();
  });
});
