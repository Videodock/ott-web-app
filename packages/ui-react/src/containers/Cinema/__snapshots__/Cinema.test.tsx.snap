// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Cinema> > renders and matches snapshot 1`] = `
<div>
  <div
    class="_fade_555f3c"
    style="transition: opacity 0.25s ease-in-out; opacity: 0;"
  >
    <div
      class="_cinema_555f3c"
    >
      <div
        class="_loadingOverlay_eb61cb _inline_eb61cb"
      >
        <div
          class="_buffer_d122df"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
      <div
        style="transition: opacity 0.25s ease-in-out; opacity: 0;"
      >
        <div
          class="_playerOverlay_555f3c"
        >
          <div
            class="_playerContent_555f3c"
          >
            <div
              aria-label="common:back"
              class="_iconButton_0fef65 _backButton_555f3c"
              role="button"
              tabindex="0"
            >
              <svg
                class="_icon_585b29"
                height="48px"
                viewBox="0 0 48 48"
                width="48px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M24 5A19 19 0 1 0 24 43A19 19 0 1 0 24 5Z"
                  fill="#039be5"
                />
                <path
                  d="M26.572,29.036h4.917l0.772-4.995h-5.69v-2.73c0-2.075,0.678-3.915,2.619-3.915h3.119v-4.359c-0.548-0.074-1.707-0.236-3.897-0.236c-4.573,0-7.254,2.415-7.254,7.917v3.323h-4.701v4.995h4.701v13.729C22.089,42.905,23.032,43,24,43c0.875,0,1.729-0.08,2.572-0.194V29.036z"
                  fill="#fff"
                />
              </svg>
            </div>
            <div>
              <h1
                class="_title_555f3c"
              >
                Test item title
              </h1>
              <div
                class="_metaContainer_555f3c"
              >
                <div
                  class="_primaryMetadata_555f3c"
                >
                  Primary metadata
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
